### PR TITLE
SHAKE128 / SHAKE256 support

### PR DESCRIPTION
**Pull request by vincenthz**
supersede #131

### CODE REVIEW

File: Crypto/Hash/SHAKE.hs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    hashInternalFinalize      = shakeFinalizeOutput (Proxy :: Proxy bitLen)
+
+shakeFinalizeOutput :: (IsDivisibleBy8 bitLen, KnownNat bitLen)
+                    => proxy bitLen
+                    -> Ptr (Context a)
+                    -> Ptr (Digest a)
+                    -> IO ()
+shakeFinalizeOutput d ctx dig = do
+    c_sha3_finalize_shake ctx
+    c_sha3_output ctx dig (fromInteger (natVal d `div` 8))
```
Comment (ocheron) [inline Crypto/Hash/SHAKE.hs:76]: Any reason for not using function `byteLen` ?

File: Crypto/Hash/SHA3.hs [hunk 2]
```diff
@@ -6,7 +6,7 @@
 -- Portability : unknown
 --
 -- module containing the binding functions to work with the
--- SHA3 cryptographic hash.
+-- SHA3 cryptographic hash and extendable output functions.
```
Comment (ocheron) [inline Crypto/Hash/SHA3.hs]: This file should be reverted entirely.

File: Crypto/Hash/SHAKE.hs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+-- | SHAKE128 (128 bits) extendable output function.  Supports an arbitrary
+-- digest size (multiple of 8 bits), to be specified as a type parameter
+-- of kind 'Nat'.
+--
+-- Note: outputs from @'SHAKE128' n@ and @'SHAKE128' m@ for the same input are
+-- correlated (one being a prefix of the other).  Results are unrelated to
+-- 'SHAKE256' results.
+data SHAKE128 (bitlen :: Nat) = SHAKE128
+    deriving (Show, Typeable)
```
Comment (ocheron) [inline Crypto/Hash/SHAKE.hs:41]: Deriving `Data` is missing for SHAKE128 and SHAKE256

**Comment by ocheron** [comment]
OK for using a separate module and conditional compilation. The latter will be a necessary evil when it comes to dependencies from other packages. I guess this situation will be only "temporary".

About generalization of SHA3 output length to any Nat: only specific sizes have been standardized, so I think it would be more confusing than useful. For Stein / Keccak: why not.

About what happens when the bit length is not divisible by 8: I don't view it so problematic. FIPS 202 defines an encoding for any bit size, even when not aligned on full bytes. In the examples there is [one with partial bytes](http://csrc.nist.gov/groups/ST/toolkit/documents/Examples/ShakeTruncation.pdf). I just did not take time to implement it. It's an optional feature and I think it will much less popular.

So I have no objection if you want to keep the `IsDivisibleBy8` type constraint, it's nicely done. But the justification for all the language extensions involved is rather weak at this point. I can implement the truncation if you prefer instead.

Btw I'm also a bit worried about the documentation in `GHC.TypeLits`, explaining the module should not be used directly but through a library instead. This library does not seem to exist yet.

In crypto code we have lots of bytearrays with length known at compile time, so there is probably room for a library providing necessary utilities. Something like constraints with more expressivity than ByteArray / ByteArrayAccess.

**Comment by vincenthz** [comment]
I haven't looked at FIPS 202 bit size encoding, but if the scheme is defined at the bit level it make sense to put on the constraint on the bit size that are actually supported right now. if/once we get the bitsize support, then we can remove the constraint.

Related to the extensions:

* I don't see it too much of an issue going forward, the needed extensions are actually not too bad, mostly TypeFamilies and DataKind, some of them are somewhat optional like TypeError and ConstraintKind.
* We could support IsDivisibleBy8 as a dummy constraint for compiler that doesn't have the needed extension. it wouldn't actually provide the real constraint, but at least you'll be able to write the constraint for future compat. Maybe a moot issue provided that older compiler don't support `Nat` anyway

I've started thinking about adding sized vector in `foundation` already, so maybe this would be useful there. I'm actually thinking of adding the `IsDiv8` constraint in foundation, and some other related things to what I played with coming up with IsDivisibleBy8.

**Comment by ocheron** [comment]
OK, thanks. That will be an example of what can be done.

**Comment by ocheron** [comment]
For information I pushed a modification directly to master so that the test suite runs with 7.6.3.
A bit too late for the new release unfortunately.

**Comment by vincenthz** [comment]
@ocheron if you want we can do a quick .1 release to fix this issue ?

**Comment by ocheron** [comment]
It's not like I need it, but if you think a new relase will be useful for someone, then yes of course.
What are the steps?

**Pull request merged**